RecoHGCal/TICL/plugins/HGCDoublet.h

```diff
@@ -0,0 +1,176 @@
+      for (int j = 0; j < vs; ++j) {
+        auto otherDoubletId = innerDoublets[i + j];
+        auto &otherDoublet = allDoublets[otherDoubletId];
+        if (ok[j]) {
+          otherDoublet.tagAsOuterNeighbor(doubletId);
+          allDoublets[doubletId].tagAsInnerNeighbor(otherDoubletId);
+        }
+      }
+    };
+    auto lim = VSIZE * (nDoublets / VSIZE);
+    for (int i = 0; i < lim; i += VSIZE) loop(i, VSIZE);
+    loop(lim, nDoublets - lim);
+
+    if (debug) {
+      std::cout << "Found " << theInnerNeighbors_.size() << " compatible doublets out of "
+                << nDoublets << " considered" << std::endl;
+    }
+    return theInnerNeighbors_.empty();
+  }
+
+  int areAligned(double xi, double yi, double zi, double xo, double yo, double zo,
+                 float minCosTheta, float minCosPointing, bool debug = false) {
+    auto dx1 = xo - xi;
+    auto dy1 = yo - yi;
+    auto dz1 = zo - zi;
+
+    auto dx2 = getInnerX() - xi;
+    auto dy2 = getInnerY() - yi;
+    auto dz2 = getInnerZ() - zi;
+
+    // inner product
+    auto dot = dx1 * dx2 + dy1 * dy2 + dz1 * dz2;
+    // magnitudes
+    auto mag1 = std::sqrt(dx1 * dx1 + dy1 * dy1 + dz1 * dz1);
+    auto mag2 = std::sqrt(dx2 * dx2 + dy2 * dy2 + dz2 * dz2);
+    // angle between the vectors
+    auto cosTheta = dot / (mag1 * mag2);
+    if (debug) {
+      std::cout << "dot: " << dot << " mag1: " << mag1 << " mag2: " << mag2
+                << " cosTheta: " << cosTheta << " isWithinLimits: " << (cosTheta > minCosTheta)
+                << std::endl;
+    }
+
+    // Now check the compatibility with the pointing origin.
+    // TODO(rovere): pass in also the origin, which is now fixed at (0,0,0)
+    // The compatibility is checked only for the innermost doublets: the
+    // one with the outer doublets comes in by the alignment requirement of
+    // the doublets themeselves
+    auto dot_pointing = dx2 * xi + dy2 * yi + dz2 * zi;
+    auto mag_pointing = std::sqrt(xi * xi + yi * yi + zi * zi);
+    auto cosTheta_pointing = dot_pointing / (mag2 * mag_pointing);
+    if (debug) {
+      std::cout << "dot_pointing: " << dot_pointing << " mag_pointing: " << mag_pointing
+                << " mag2: " << mag2 << " cosTheta_pointing: " << cosTheta_pointing
+                << " isWithinLimits: " << (cosTheta_pointing < minCosPointing) << std::endl;
+    }
+
+    return (cosTheta > minCosTheta) && (cosTheta_pointing > minCosPointing);
+  }
+
+  void findNtuplets(std::vector<HGCDoublet> &allDoublets, HGCntuplet &tmpNtuplet) {
+    if (!alreadyVisited_) {
+      alreadyVisited_ = true;
+      tmpNtuplet.push_back(theDoubletId_);
+      unsigned int numberOfOuterNeighbors = theOuterNeighbors_.size();
+      for (unsigned int i = 0; i < numberOfOuterNeighbors; ++i) {
+        allDoublets[theOuterNeighbors_[i]].findNtuplets(allDoublets, tmpNtuplet);
+      }
+    }
+  }
+
+ private:
+  const std::vector<reco::CaloCluster> *layerClusters_;
+  std::vector<int> theOuterNeighbors_;
```

Comment:
don't need both "the" and trailing "_" to denote a member variable (just use the latter) (for all members below as well)